docs/source/prgpatterns/lockstep.rst

```diff
@@ -19,21 +19,118 @@ Code which is implemented by the *lockstep programming model* is free of any dep
 To simplify the implementation, each index within a domain can be seen as a *virtual worker* which is processing one data element (like the common workflow to programming CUDA).
 Each *worker* :math:`i` can be executed as :math:`N_i` *virtual workers* (:math:`1:N_i`).
 
+To transfer information from a subsequent lockstep to another you can use a context variable ``lockstep::Variable``, similar to a temporary local variable in a function.
+
+Functors passed into lockstep routines can have three different parameter signatures.
+
+* No parameter, if the work is not requiring the linear index within a domain: ``[&](){ }``
+
+
+* An unsigned 32bit integral parameter if the work depends on indices within the domain ``range [0,domain size)``: ``[&](uint32_t const linearIdx){}``
+
+
+* ``lockstep::Idx`` as parameter. lockstep::Idx is holing the linear index within the domain and meta information to access a context variables: ``[&](pmacc::mappings::threads::lockstep::Idx const idx){}``
+
+
 pmacc helpers
 -------------
 
-.. doxygenstruct:: pmacc::mappings::threads::IdxConfig
+.. doxygenstruct:: pmacc::lockstep::Config
    :project: PIConGPU
 
-.. doxygenstruct:: pmacc::memory::CtxArray
+.. doxygenstruct:: pmacc::lockstep::Idx
+  :project: PIConGPU
+
+.. doxygenstruct:: pmacc::lockstep::Worker
+  :project: PIConGPU
+
+.. doxygenstruct:: pmacc::lockstep::Variable
    :project: PIConGPU
 
-.. doxygenstruct:: pmacc::mappings::threads::ForEachIdx
+.. doxygenstruct:: pmacc::lockstep::ForEach
    :project: PIConGPU
 
 Common Patterns
 ---------------
 
+Create a Context Variable
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+* ... and initialize with the index of the virtual worker
+
+.. code-block:: cpp
+
+    uint32_t const workerIdx = cupla::threadIdx(acc).x;
+    constexpr uint32_t frameSize = 256;
+    constexpr uint32_t numWorkers = 42;
+    auto forEachParticleSlotInFrame = lockstep::makeForEach<frameSize, numWorkers>(workerIdx);
+    auto vIdx = forEachParticleSlotInFrame(
+        [](lockstep::Idx const idx) -> int32_t
+        {
+            return idx;
+        }
+    );
+
+    // is equal to
+
+    uint32_t const workerIdx = cupla::threadIdx(acc).x;
+    constexpr uint32_t frameSize = 256;
+    constexpr uint32_t numWorkers = 42;
+    auto forEachParticleSlotInFrame = lockstep::makeForEach<frameSize, numWorkers>(workerIdx);
+    // variable will be uninitialized
+    auto vIdx = lockstep::makeVar<int32_t>(forEachParticleSlotInFrame);
+    forEachParticleSlotInFrame(
+        [&](lockstep::Idx const idx)
```

Comment:
Could `lockstep::Idx` be `auto` instead? Asking both as a technical possibility, and if that is a good recommended way

Reply:
Yes could be auto too but I prefer to be explicit to see the difference between `uint32_t const` which can be used too.
`lockstep::Idx` is also better for search an replace^^